Bug: Migration is failed on v3.6.0-RC when authentication to MongoDB is enabled

Environment
------------

### Host

| item     | version |
| ---      | --- |
|OS        | Debian 10 (Busteer)|
|GROWI     |3.6.0-RC|
|node.js   |v10.17.0 and v12.13.0|
|npm       |6.11.3|
|Using Docker|no|
|Using [growi-docker-compose][growi-docker-compose]|no|

[growi-docker-compose]: https://github.com/weseek/growi-docker-compose

*(Accessing https://{GROWI_HOST}/admin helps you to fill in above versions)*



How to reproduce? (再現手順)
---------------------------

1. yarn && npm run build:prod && npm run server:prod

What happens? (症状)
---------------------

- Migration でエラー

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'run', 'server:prod' ]
2 info using npm@6.12.0
3 info using node@v12.13.0
4 verbose run-script [ 'preserver:prod', 'server:prod', 'postserver:prod' ]
5 info lifecycle growi@3.6.0-RC~preserver:prod: growi@3.6.0-RC
6 verbose lifecycle growi@3.6.0-RC~preserver:prod: unsafe-perm in lifecycle true
7 verbose lifecycle growi@3.6.0-RC~preserver:prod: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/srv/growi/growi/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
8 verbose lifecycle growi@3.6.0-RC~preserver:prod: CWD: /srv/growi/growi
9 silly lifecycle growi@3.6.0-RC~preserver:prod: Args: [ '-c', 'npm run migrate' ]
10 silly lifecycle growi@3.6.0-RC~preserver:prod: Returned: code: 1  signal: null
11 info lifecycle growi@3.6.0-RC~preserver:prod: Failed to exec preserver:prod script
12 verbose stack Error: growi@3.6.0-RC preserver:prod: `npm run migrate`
12 verbose stack Exit status 1
12 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
12 verbose stack     at EventEmitter.emit (events.js:210:5)
12 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
12 verbose stack     at ChildProcess.emit (events.js:210:5)
12 verbose stack     at maybeClose (internal/child_process.js:1021:16)
12 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
13 verbose pkgid growi@3.6.0-RC
14 verbose cwd /srv/growi/growi
15 verbose Linux 4.19.0-6-amd64
16 verbose argv "/usr/bin/node" "/usr/bin/npm" "run" "server:prod"
17 verbose node v12.13.0
18 verbose npm  v6.12.0
19 error code ELIFECYCLE
20 error errno 1
21 error growi@3.6.0-RC preserver:prod: `npm run migrate`
21 error Exit status 1
22 error Failed at the growi@3.6.0-RC preserver:prod script.
22 error This is probably not a problem with npm. There is likely additional logging output above.
23 verbose exit [ 1, true ]
```



What is the expected result? (期待される動作)
-------------------------------------------

- 正常起動
- 



Note
----

- よろしくお願いします

commit d63c5fbe5342829c67724d10e6613a0e675b7459 になるようです。
ログは、npm のエラーログです。コマンドは、bash からコマンドラインに入力して実行しています。今試したところ、最後の `npm run server:prod` で以下の出力を得ました。

```
> growi@3.6.0-RC preserver:prod /srv/growi/growi
> npm run migrate


> growi@3.6.0-RC migrate /srv/growi/growi
> npm run migrate:up


> growi@3.6.0-RC migrate:up /srv/growi/growi
> migrate-mongo up -f config/migrate.js

(node:21986) DeprecationWarning: current Server Discovery and Monitoring engine is deprecated, and will be removed in a future version. To use the new Server Discover and Monitoring engine, pass option { useUnifiedTopology: true } to the MongoClient constructor.
ERROR: command find requires authentication
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! growi@3.6.0-RC migrate:up: `migrate-mongo up -f config/migrate.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the growi@3.6.0-RC migrate:up script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /srv/growi/.npm/_logs/2019-12-09T14_24_39_306Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! growi@3.6.0-RC migrate: `npm run migrate:up`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the growi@3.6.0-RC migrate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /srv/growi/.npm/_logs/2019-12-09T14_24_39_372Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! growi@3.6.0-RC preserver:prod: `npm run migrate`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the growi@3.6.0-RC preserver:prod script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /srv/growi/.npm/_logs/2019-12-09T14_24_39_430Z-debug.log
```

--- a/config/migrate.js
+++ b/config/migrate.js
@@ -16,8 +16,12 @@ const mongoUri = getMongoUri();
 // parse url
 const url = new URL(mongoUri);
 
+const authStr = (url.username.length > 0 && url.password.length > 0)
+  ? `${url.username}:${url.password}@`
+  : '';
+
 const mongodb = {
-  url: `${url.protocol}//${url.host}`,
+  url: `${url.protocol}//${authStr}${url.host}${url.search}`,
   databaseName: url.pathname.substring(1), // omit heading slash
   options: {
     useNewUrlParser: true, // removes a deprecation warning when connecting

--- a/src/test/config/migrate.test.js
+++ b/src/test/config/migrate.test.js
@@ -1,0 +1,37 @@
+describe('config/migrate.js', () => {
+
+  beforeEach(async(done) => {
+    jest.resetModules();
+    done();
+  });
+
+  /* eslint-disable indent */
+  describe.each`
+    MONGO_URI                                         | expectedUrl                                     | expectedDbName
+    ${'mongodb://example.com/growi'}                  | ${'mongodb://example.com'}                      | ${'growi'}
+    ${'mongodb://user:pass@example.com/growi'}        | ${'mongodb://user:pass@example.com'}            | ${'growi'}
+    ${'mongodb://example.com/growi?replicaSet=mySet'} | ${'mongodb://example.com?replicaSet=mySet'}     | ${'growi'}
+  `('returns', ({ MONGO_URI, expectedUrl, expectedDbName }) => {
+    test(`when 'MONGO_URI' is '${MONGO_URI}`, () => {
+
+      // mock for mongoose-utils
+      jest.doMock('@commons/util/mongoose-utils', () => {
+        return {
+          getMongoUri: () => {
+            return MONGO_URI;
+          },
+        };
+      });
+
+      const { mongoUri, mongodb } = require('@root/config/migrate');
+
+      jest.dontMock('@commons/util/mongoose-utils');
+
+      expect(mongoUri).toBe(MONGO_URI);
+      expect(mongodb.url).toBe(expectedUrl);
+      expect(mongodb.databaseName).toBe(expectedDbName);
+    });
+  });
+  /* eslint-enable indent */
+
+});

